ext/mbstring/libmbfl/mbfl/mbfilter.c

```diff
@@ -245,12 +245,12 @@ static int mbfl_estimate_encoding_likelihood(int c, void *void_data)
 	 * it's the wrong one. */
 	if (c == MBFL_BAD_INPUT) {
 		data->num_illegalchars++;
-	} else if (php_unicode_is_cntrl(c) || php_unicode_is_private(c)) {
+	} else if (c < 0x9 || (c >= 0xE && c <= 0x1F) || (c >= 0xE000 && c <= 0xF8FF) || c >= 0xF0000) {
 		/* Otherwise, count how many control characters and 'private use'
 		 * codepoints we see. Those are rarely used and may indicate that
 		 * the candidate encoding is not the right one. */
 		data->score += 10;
-	} else if (php_unicode_is_punct(c)) {
+	} else if ((c >= 0x21 && c <= 0x2F) || (c >= 0x3A && c <= 0x40) || (c >= 0x2000 && c <= 0x206F)) {
```

Comment:
These character ranges aren't equivalent to the previous check -- what's the rationale for checking these ranges, but not others?

Reply:
First of all, the rationale for checking for punctuation in the first place...

Some legacy encodings deliberately encode multi-byte characters using byte values which are in the ASCII range (like all the ISO-2022 encodings, HZ, etc). If such an encoding is misinterpreted as ASCII, then typically the mangled string will have a much higher density of punctuation characters than would normally be the case. Hence, our main interest here is in punctuation characters which are present in ASCII.

I've added another range of codepoints which is also used for punctuation, but not for any reason in particular. And we could certainly add still more ranges, though I don't know if it will really help anything. In any case, I do want to keep the number of ranges checked reasonably small, for performance.

Remember these are just heuristics. If we can find other heuristics which are fast, and which make good guesses most of the time, that would be great.

At the end of the day, what we *really* need is a good corpus of realistic example texts; then we could benchmark various candidate heuristics for detection accuracy and speed, and pick the best combination based on actual data.

Reply:
Okay, that makes sense. I think it would be useful to add that explanation as a comment. Maybe we can drop the non-ascii range if we don't have an immediate use for it.